File: apps/menu-service/src/main/java/com/agentic/restaurant/menu/application/AuthValidationClient.java
package com.agentic.restaurant.menu.application;

import java.util.Map;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestTemplate;

@Component
public class AuthValidationClient {

    private final RestTemplate restTemplate;
    private final String usersServiceBaseUrl;
    private final String usersServiceToken;

    public AuthValidationClient(
        @Value("${app.auth.users-service-base-url}") String usersServiceBaseUrl,
        @Value("${app.auth.users-service-token}") String usersServiceToken
    ) {
        this.restTemplate = new RestTemplate();
        this.usersServiceBaseUrl = usersServiceBaseUrl;
        this.usersServiceToken = usersServiceToken;
    }

    public boolean validateBearerToken(String bearerToken) {
        return validateToken(bearerToken).valid();
    }

    @SuppressWarnings("unchecked")
    public TokenValidationResult validateToken(String bearerToken) {
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        headers.set("X-Service-Token", usersServiceToken);
        HttpEntity<Map<String, String>> request = new HttpEntity<>(Map.of("token", bearerToken), headers);

        try {
            ResponseEntity<Map> response = restTemplate.postForEntity(
                usersServiceBaseUrl + "/api/v1/internal/auth/validate",
                request,
                Map.class
            );
            Map<String, Object> body = response.getBody();
            if (body == null) {
                return TokenValidationResult.INVALID;
            }
            boolean valid = Boolean.TRUE.equals(body.get("valid"));
            String clientType = (String) body.get("clientType");
            String displayName = (String) body.get("displayName");
            return new TokenValidationResult(valid, clientType, displayName);
        } catch (Exception ex) {
            return TokenValidationResult.INVALID;
        }
    }

    public record TokenValidationResult(boolean valid, String clientType, String displayName) {
        static final TokenValidationResult INVALID = new TokenValidationResult(false, null, null);
    }
}

File: apps/orders-service/src/main/kotlin/com/agentic/restaurant/orders/clients/AuthValidationClient.kt
package com.agentic.restaurant.orders.clients

import org.springframework.beans.factory.annotation.Value
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType
import org.springframework.stereotype.Component
import org.springframework.web.client.RestTemplate

@Component
class AuthValidationClient(
    @Value("\${app.auth.users-service-base-url}") private val usersServiceBaseUrl: String,
    @Value("\${app.auth.users-service-token}") private val usersServiceToken: String,
) {
    private val restTemplate = RestTemplate()

    fun validateBearerToken(bearerToken: String): TokenValidationResult {
        val headers = HttpHeaders()
        headers.contentType = MediaType.APPLICATION_JSON
        headers["X-Service-Token"] = usersServiceToken

        val request = HttpEntity(mapOf("token" to bearerToken), headers)

        return try {
            val response = restTemplate.postForEntity(
                "$usersServiceBaseUrl/api/v1/internal/auth/validate",
                request,
                Map::class.java,
            )
            val body = response.body ?: return TokenValidationResult(valid = false)
            val valid = body["valid"] as? Boolean ?: false
            val userId = (body["userId"] as? Number)?.toLong()
            val clientType = body["clientType"] as? String
            val displayName = body["displayName"] as? String
            TokenValidationResult(valid = valid, userId = userId, clientType = clientType, displayName = displayName)
        } catch (_: Exception) {
            TokenValidationResult(valid = false)
        }
    }
}
